### PR TITLE
PE-4583: Video Preview (In-app)

### PR DESCRIPTION
**Pull request by kunstmusik**
Video player w/full screen view. 

--- Releases ---
Android release: https://appdistribution.firebase.google.com/testerapps/1:305132849030:android:6cf0cd5ec064fad3ffce07/releases/0prio1hor6ls0

### CODE REVIEW

File: assets/config/prod.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
   "enableQuickSyncAuthoring": true,
   "enableMultipleFileDownload": true,
   "enableVideoPreview": false,
+  "enableAudioPreview": false,
```
Comment (thiagocarvalhodev) [inline assets/config/prod.json:11]: We will need to update it to true when releasing

File: assets/config/staging.json [hunk 2]
```diff
@@ -8,6 +8,7 @@
   "enableQuickSyncAuthoring": true,
   "enableMultipleFileDownload": true,
   "enableVideoPreview": false,
+  "enableAudioPreview": false,
```
Comment (thiagocarvalhodev) [inline assets/config/staging.json:11]: The same for the previous comment

File: lib/blocs/fs_entry_preview/fs_entry_preview_cubit.dart [hunk 3]
```diff
@@ -175,6 +184,16 @@ class FsEntryPreviewCubit extends Cubit<FsEntryPreviewState> {
               case 'image':
                 emitImagePreview(file, previewUrl);
                 break;
+
+              case 'audio':
```
Comment (thiagocarvalhodev) [inline lib/blocs/fs_entry_preview/fs_entry_preview_cubit.dart:188]: @kunstmusik , it seems this branch is out to date. Let's sync with dev

File: lib/dev_tools/app_dev_tools.dart [hunk 4]
```diff
@@ -233,6 +233,19 @@ class AppConfigWindowManagerState extends State<AppConfigWindowManager> {
       type: ArDriveDevToolOptionType.bool,
     );
 
+    ArDriveDevToolOption enableAudioPreviewOption = ArDriveDevToolOption(
```
Comment (thiagocarvalhodev) [inline lib/dev_tools/app_dev_tools.dart:236]: 🚀

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 5]
```diff
@@ -1,5 +1,7 @@
 part of '../drive_detail_page.dart';
 
+const List<double> _speedOptions = [.25, .5, .75, 1, 1.25, 1.5, 1.75, 2];
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:3]: Should we use the `normal` option here, as well?

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 6]
```diff
@@ -35,86 +37,1398 @@ class _FsEntryPreviewWidgetState extends State<FsEntryPreviewWidget> {
...
+    logger.d('Disposing video player');
+    _videoPlayerController.removeListener(_listener);
+    _videoPlayerController.dispose();
+    super.dispose();
+  }
+
+  void _listener() {
+    setState(() {
+      if (_videoPlayerController.value.hasError) {
+        logger.d('>>> ${_videoPlayerController.value.errorDescription}');
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:123]: nit: we could use the `logger.e`

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 7]
```diff
@@ -35,86 +37,1398 @@ class _FsEntryPreviewWidgetState extends State<FsEntryPreviewWidget> {
...
+            VideoPlayer(_videoPlayerController, key: const Key('videoPlayer'));
+      }
+    });
+  }
+
+  void goFullScreen() {
+    bool wasPlaying = _videoPlayerController.value.isPlaying;
+    if (wasPlaying) {
+      _videoPlayerController.pause().catchError((error) {
+        logger.d('Error pausing video: $error');
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:144]: nit: use `logger.e`

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 8]
```diff
@@ -35,86 +37,1398 @@ class _FsEntryPreviewWidgetState extends State<FsEntryPreviewWidget> {
...
+                initialPosition: _videoPlayerController.value.position,
+                initialIsPlaying: wasPlaying,
+                initialVolume: _videoPlayerController.value.volume,
+                onClose: (position, isPlaying, volume) async {
+                  _videoPlayerController.seekTo(position);
+                  _videoPlayerController.setVolume(volume);
+                  if (isPlaying) {
+                    await _lock.synchronized(() async {
+                      await _videoPlayerController.play().catchError((e) {
+                        logger.d('Error playing video: $e');
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:166]: nit: use `logger.e`

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 9]
```diff
@@ -35,86 +37,1398 @@ class _FsEntryPreviewWidgetState extends State<FsEntryPreviewWidget> {
...
+
+    return VisibilityDetector(
+        key: const Key('video-player'),
+        onVisibilityChanged: (VisibilityInfo info) async {
+          if (mounted) {
+            if (info.visibleFraction < 0.5 &&
+                _videoPlayerController.value.isPlaying) {
+              await _lock.synchronized(() async {
+                await _videoPlayerController.pause().catchError((error) {
+                  logger.d('Error pausing video: $error');
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:194]: nit: use the `logger.e`

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 10]
```diff
@@ -35,86 +37,1398 @@ class _FsEntryPreviewWidgetState extends State<FsEntryPreviewWidget> {
...
+                      });
+                    }
+                  });
+                }
+              });
+            },
+            child: Container(color: Colors.black.withOpacity(0.0)),
+          ),
+        ),
+        AnimatedOpacity(
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:627]: We can improve the readability of this widget by splitting its sub-widgets into private method or classes.

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 11]
```diff
@@ -35,86 +37,1398 @@ class _FsEntryPreviewWidgetState extends State<FsEntryPreviewWidget> {
...
+
+        _videoPlayerController =
+            VideoPlayerController.networkUrl(Uri.parse(widget.videoUrl));
+        _videoPlayerController.initialize();
+        _videoPlayerController.addListener(_listener);
+
+        _videoPlayer =
+            VideoPlayer(_videoPlayerController, key: const Key('videoPlayer'));
+      }
+    });
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:538]: We have almost the same logic on the `VideoPlayerWidget`. We could implement an abstraction for that, so we don't need to repeat the code. I'm not requesting for now, we can address it in a tech debt in the future.

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart [hunk 12]
```diff
@@ -35,86 +37,1398 @@ class _FsEntryPreviewWidgetState extends State<FsEntryPreviewWidget> {
...
+          });
+        } else {
+          _videoPlayer = VideoPlayer(_videoPlayerController,
+              key: const Key('videoPlayer'));
+        }
+      });
+    });
+    _videoPlayerController.addListener(_listener);
+
+    SystemChrome.setPreferredOrientations([
```
Comment (thiagocarvalhodev) [inline lib/pages/drive_detail/components/fs_entry_preview_widget.dart:507]: 🚀

**Review by thiagocarvalhodev — APPROVED** [review]
The feature is looking GREAT! I just left a few comments about using the `logger.e` method when logging errors. We have some duplicated code that we should address in a tech debt, but not for now.
LGTM

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart (hunk 5 above)
Comment (kunstmusik) [reply lib/pages/drive_detail/components/fs_entry_preview_widget.dart]: I'm using the numbers to set the speed; could use 'normal' here but it it wouldn't make the code that uses it much different (would have to check for 'normal' and use 1 for speed param; current code checks for 1 and uses 'normal' for speed label).

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart (hunk 6 above)
Comment (kunstmusik) [reply lib/pages/drive_detail/components/fs_entry_preview_widget.dart:123]: Did a pass through and changed all error message logging to use logger.e

File: lib/pages/drive_detail/components/fs_entry_preview_widget.dart (hunk 11 above)
Comment (kunstmusik) [reply lib/pages/drive_detail/components/fs_entry_preview_widget.dart]: Added PE-4697 tech debt task to look at refactoring for clarity and code reuse.

File: pubspec.lock
Comment (matibat) [inline pubspec.lock:1]: Did we intend to update these versions?

**Review by matibat — APPROVED** [review]
LGTM

**Comment by thiagocarvalhodev** [comment]
Closing this one. The changes here were merged on this PR: #1425

**Pull request merged**